File: fields/src/main/java/network/aika/type/Type.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package network.aika.type;

import network.aika.fields.defs.FieldDefinition;
import network.aika.type.relations.Relation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

import static network.aika.type.FlattenedType.createInputFlattenedType;
import static network.aika.type.FlattenedType.createOutputFlattenedType;

/**
 * @author Lukas Molzberger
 */
public class Type {

    protected static final Logger LOG = LoggerFactory.getLogger(Type.class);

    public static final Comparator<Type> TYPE_COMPARATOR = Comparator.
            comparingInt(Type::getDepth)
            .thenComparing(Type::getId);

    private final short id;
    private final String name;

    protected final List<Type> parents = new ArrayList<>();
    protected final List<Type> children = new ArrayList<>();

    private final TypeRegistry registry;

    private final Set<FieldDefinition> fieldDefinitions = new HashSet<>();

    protected List<Relation> relations = new ArrayList<>();

    private Integer depth;

    private FlattenedType flattenedTypeInputSide;
    private FlattenedType flattenedTypeOutputSide;

    public Type(TypeRegistry registry, String name) {
        this.name = name;
        this.registry = registry;

        id = registry.register(this);
    }

    public short getId() {
        return id;
    }

    public boolean isAbstract() {
        return !children.isEmpty();
    }


    public Relation[] getRelations() {
        return relations.toArray(new Relation[0]);
    }

    public void initFlattenedType() {
        Set<FieldDefinition> fieldDefs = getCollectFlattenedFieldDefinitions();

        flattenedTypeInputSide = createInputFlattenedType(this, fieldDefs);
        flattenedTypeOutputSide = createOutputFlattenedType(this, fieldDefs, flattenedTypeInputSide);
    }

    public Set<FieldDefinition> getCollectFlattenedFieldDefinitions() {
        return collectTypes()
                .stream()
                .flatMap(t -> t.getFieldDefinitions().stream())
                .collect(Collectors.toSet());
    }

    public SortedSet<Type> collectTypes() {
        TreeSet<Type> sortedTypes = new TreeSet<>(TYPE_COMPARATOR);
        collectTypesRecursiveStep(sortedTypes);
        return sortedTypes;
    }

    public void collectTypesRecursiveStep(SortedSet<Type> sortedTypes) {
        parents.forEach(p ->
                p.collectTypesRecursiveStep(sortedTypes)
        );
        sortedTypes.add(this);
    }

    public Integer getDepth() {
        if(depth == null)
            depth = parents.stream()
                .mapToInt(Type::getDepth)
                .max()
                .orElse(0);

        return depth;
    }

    public boolean isInstanceOf(Obj obj) {
        return isInstanceOf(obj.getType());
    }

    public boolean isInstanceOf(Type type) {
        return this == type ||
                parents.stream().anyMatch(p ->
                        p.isInstanceOf(type)
                );
    }

    public String getName() {
        return name;
    }

    public TypeRegistry getTypeRegistry() {
        return registry;
    }

    public FlattenedType getFlattenedTypeInputSide() {
        if(flattenedTypeInputSide == null)
            throw new RuntimeException("Type has not been flattened yet. TypeRegistry.flattenTypeHierarchy() needs to be called beforehand.");

        return flattenedTypeInputSide;
    }

    public FlattenedType getFlattenedTypeOutputSide() {
        if(flattenedTypeOutputSide == null)
            throw new RuntimeException("Type has not been flattened yet. TypeRegistry.flattenTypeHierarchy() needs to be called beforehand.");

        return flattenedTypeOutputSide;
    }

    public void setFieldDefinition(FieldDefinition fieldDef) {
        fieldDef.setFieldId(registry.createFieldId());
        fieldDefinitions.add(fieldDef);
    }

    public Set<FieldDefinition> getFieldDefinitions() {
        return fieldDefinitions;
    }

    public <T extends Type> Type addParent(T p) {
        parents.add(p);
        p.children.add(this);

        return this;
    }

    public List<? extends Type> getParents() {
        return parents;
    }

    public List<? extends Type> getChildren() {
        return children;
    }

    protected <R> R getFromParent(Function<Type, R> f) {
        return parents.stream()
                .map(f)
                .filter(Objects::nonNull)
                .findFirst()
                .orElse(null);
    }

    public String toString() {
        return name;
    }
}

File: src/fields/type.cpp
#include "fields/type.h"
#include "fields/obj.h"
#include "fields/type_registry.h"
#include "fields/field_definition.h"
#include "fields/flattened_type.h"
#include <iostream>
#include <set>
#include <algorithm>
#include <stdexcept>

bool Type::TypeComparator::operator()(const Type *t1, const Type *t2) const {
    if (t1->getDepth() != t2->getDepth()) return t1->getDepth() < t2->getDepth();
    return t1->getId() < t2->getId();
};

Type::Type(TypeRegistry* registry, const std::string& name)
    : registry(registry), name(name), depth(std::nullopt) {
    id = registry->registerType(this);
}

short Type::getId() const {
    return id;
}

bool Type::isAbstract() const {
    return !children.empty();
}

std::vector<Relation*> Type::getRelations() const {
    return relations;
}

void Type::initFlattenedType() {
    auto fieldDefs = getCollectFlattenedFieldDefinitions();

    flattenedTypeInputSide = FlattenedType::createInputFlattenedType(this, fieldDefs);
    flattenedTypeOutputSide = FlattenedType::createOutputFlattenedType(this, fieldDefs, flattenedTypeInputSide);
}

std::set<FieldDefinition*> Type::getCollectFlattenedFieldDefinitions() {
    std::set<FieldDefinition*> fieldDefs;
    for (const auto& t : collectTypes()) {
        auto tFieldDefs = t->getFieldDefinitions();
        fieldDefs.insert(tFieldDefs.begin(), tFieldDefs.end());
    }
    return fieldDefs;
}

std::set<Type*> Type::collectTypes() {
    std::set<Type*> sortedTypes;
    collectTypesRecursiveStep(sortedTypes);
    return sortedTypes;
}

void Type::collectTypesRecursiveStep(std::set<Type*>& sortedTypes) {
    for (const auto& p : parents) {
        p->collectTypesRecursiveStep(sortedTypes);
    }
    sortedTypes.insert(this);
}

void Type::initDepth() {
    if (!depth.has_value()) {
        depth = 0;
        for (const auto& p : parents) {
            depth = std::max(depth.value(), p->getDepth() + 1);
        }
    }
}

int Type::getDepth() const {
    return depth.value();
}

bool Type::isInstanceOf(Obj* obj) {
    return isInstanceOf(obj->getType());
}

bool Type::isInstanceOf(Type* type) {
    return id == type->id || std::any_of(parents.begin(), parents.end(), [&](const auto& p) {
        return p->isInstanceOf(type);
    });
}

std::string Type::getName() const {
    return name;
}

TypeRegistry* Type::getTypeRegistry() {
    return registry;
}

FlattenedType* Type::getFlattenedTypeInputSide() {
    if (!flattenedTypeInputSide) {
        throw std::runtime_error("Type has not been flattened yet. TypeRegistry.flattenTypeHierarchy() needs to be called beforehand.");
    }
    return flattenedTypeInputSide;
}

FlattenedType* Type::getFlattenedTypeOutputSide() {
    if (!flattenedTypeOutputSide) {
        throw std::runtime_error("Type has not been flattened yet. TypeRegistry.flattenTypeHierarchy() needs to be called beforehand.");
    }
    return flattenedTypeOutputSide;
}

void Type::setFieldDefinition(FieldDefinition* fieldDef) {
    fieldDef->setFieldId(registry->createFieldId());
    fieldDefinitions.insert(fieldDef);
}

std::set<FieldDefinition*> Type::getFieldDefinitions() const {
    return fieldDefinitions;
}

Type& Type::addParent(Type* p) {
    parents.push_back(p);
    p->children.push_back(this);
    return *this;
}

std::vector<Type*> Type::getParents() const {
    return parents;
}

std::vector<Type*> Type::getChildren() const {
    return children;
}

template <typename R>
R Type::getFromParent(std::function<R(Type*)> f) {
    for (const auto& p : parents) {
        R result = f(p);
        if (result) {
            return result;
        }
    }
    return nullptr;
}

std::string Type::toString() const {
    return name;
}

